server/src/test/java/org/elasticsearch/snapshots/SnapshotsServiceTests.java

```diff
@@ -0,0 +1,526 @@
+import static java.util.Collections.emptySet;
+import static org.elasticsearch.env.Environment.PATH_HOME_SETTING;
+import static org.elasticsearch.node.Node.NODE_NAME_SETTING;
+import static org.elasticsearch.transport.TransportService.NOOP_TRANSPORT_INTERCEPTOR;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hamcrest.Matchers.hasSize;
+import static org.mockito.Matchers.any;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class SnapshotsServiceTests extends ESTestCase {
+
+    private TestClusterNodes testClusterNodes;
+
+    @Before
+    public void createServices() {
+        // TODO: Random number of master nodes and simulate master failover states
+        testClusterNodes = new TestClusterNodes(1, randomIntBetween(2, 10));
+    }
+
+    @After
+    public void stopServices() {
+        testClusterNodes.nodes.values().forEach(
+            n -> {
+                n.indicesService.close();
+                n.clusterService.close();
+                n.indicesClusterStateService.close();
+                n.nodeEnv.close();
+            }
+        );
+    }
+
+    public void testSuccessfulSnapshot() {
+
+        String repoName = "repo";
+        String snapshotName = "snapshot";
+        final String index = "test";
+
+        final int shards = randomIntBetween(1, 10);
+        final Repository repository = createRepository();
+        testClusterNodes.nodes.values().forEach(
+            node -> when(node.repositoriesService.repository(repoName)).thenReturn(repository)
+        );
+
+        ClusterState initialClusterState =
+            new ClusterState.Builder(ClusterName.DEFAULT)
+                .nodes(testClusterNodes.randomDiscoveryNodes())
+                .metaData(
+                    MetaData.builder().putCustom(
+                        RepositoriesMetaData.TYPE,
+                        new RepositoriesMetaData(
+                            Collections.singletonList(
+                                new RepositoryMetaData(
+                                    repoName, randomAlphaOfLength(10), Settings.EMPTY
+                                )
+                            )
+                        )
+                    )
+                ).build();
+        testClusterNodes.nodes.values().forEach(testClusterNode -> testClusterNode.start(initialClusterState));
+
+        TestClusterNode masterNode = testClusterNodes.currentMaster(initialClusterState);
+        ActionFuture<CreateIndexResponse> createIndexResponseActionFuture = masterNode.client.admin().indices().create(
+            new CreateIndexRequest(index).settings(
+                Settings.builder()
+                    .put(IndexMetaData.INDEX_NUMBER_OF_SHARDS_SETTING.getKey(), shards)
+                    .put(IndexMetaData.INDEX_NUMBER_OF_REPLICAS_SETTING.getKey(), 1)
+            )
+        );
+
+        runOutstandingTasks();
```

Comment:
Running all tasks so that the index is fully set up when we create the snapshot.